source/index.ts

```diff
@@ -50,7 +50,6 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
 	*/
 	timeout?: number;
 
-	// TODO: The `throwOnTimeout` option should affect the return types of `add()` and `addAll()`
```

Comment:
Damn, as soon as I submitted the PR I realized we can't remove this just yet. The options type passed to the constructor will need to be "remembered" so we can reuse it when determining the return types for `add` and `addAll`. Because you could specify `throwOnTimeout=true` in the constructor and then only `timeout=number` when calling `add`. I'll see if I can figure out a solution...

Reply:
Okay, I couldn't find a solution so I added the todo back. I've set up [this simplified example in TS Playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAigrhBBBAJig8mYBLA9gOwGcoBeKAbyhwFsJc5gB+ALinzmoCMIAnAGioALHrgDu6fABVstek1adcuADYQAhvgFgeeHaBZsO3HlAC+AbgBQoSFADq2YIOmyGAHkw4ChAHykonnhEUABkFDR0DKzsXLwW1uDQDk5yLpHAHlhBvv7o1I6ZXkQCAOQRciV+AD7hMumscPgoEABm2PgQKPE2SY7OdanCYoXZfmTJ-a4Zgd5+YeROIuJSA1FUPAjdifZ9qavAkkOiI7P+E2lyJ0Rz4UcSFwwGLWrKhBDxlgDGymqExPCICBuADC3mAG0+wFwPBmwQgAA9gBAmv8EMg0LDiGQAeiMFlvAJUHiisQEUiUVAAApqHg4F4gsEQqEw-HXfzU2nYemgojguCQ6GYnx+ciWACQn0Z-OZAApcKzCKwOXTlAzeUzBQqxhRTABKHWWcVqNBuSQ+GU0gDmrEkAnlJIMEz2UzcRKFuptVjFxpQpvNVptdoVrHO+0OS1dGK1Hqgki9Pr9Fp41tjQZJId2DAeByOkeJoxjkigNQAbrhsCh4yazUmU7aoPbsqw3Qr-ORTFBflAWyTC8WoGWKwbTIbJbyoABHACM-g6olgaIgMt1VjHhGAVCnrHXOnwlv804AdD6ZSUAIQlFejsGTgBMs4g85xS6va43wFv2-B7X3pfLKAPW9jzQU8LwEBZ9lYKc9SsL4bwnABmB8n0XZdVxvYAEK-XdfwHf8DwQ4CUFAkpwPKNYpwERYxHuSCoGeV53ivODxwnAAWZCF0BNCWPXKg2Own8DzYoiSLIujKKEJZaKmVg+SY2C30nABWTjnxlBY7hWWT1k2V8MOUwS9wPZTRPPUjah06CryAA) that features my best attempt. I also tried adding a second generic type parameter to `add` but that didn't work either. It's unclear to me if I'm lacking the TS knowledge or if we are brushing up against the limits of TS. Maybe someone else is able to solve this; should I open a separate issue for this todo?

I still think there's value in this PR though, since it at least corrects the types for the `add` method when looked at in isolation. It just won't respect the options passed earlier to the constructor, which was already the case before this PR. However the `void` will now surface differently to before this PR which might come as a surprise to people defining constructor options:

```ts
const queue = new PQueue({throwOnTimeout: true});
const a: number = await queue.add(async () => 1, {timeout: 1});
// ❌ return type is `number | void`.
```
```ts
const queue = new PQueue({timeout: 1, throwOnTimeout: false});
const a: number | void = await queue.add(async () => 1);
// ❌ return type is `number`. there's no type error either because it's a subtype of `number | void`
```

So I guess it's a question of where you prefer the `void` type surprise to happen? Personally I think it's better to have `add` be correct in isolation and let the constructor options people deal with forcing the type to be `void`. But I can understand if you prefer `add` to incorrectly return `void` if that is a "safer" type surprise for people to deal with. Either way it's a mess!